modules/images/webp-uploads/generation.php

```diff
@@ -0,0 +1,236 @@
+<?php
+/**
+ * On demand generation functions.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Generates images on the fly for the registered image sizes and available transformations.
+ *
+ * @since n.e.x.t
+ */
+function webp_uploads_generate_missing_images() {
+	if ( ! is_404() ) {
+		return;
+	}
+
+	$uploads_dir = wp_upload_dir();
+	$baseurl     = wp_parse_url( $uploads_dir['baseurl'], PHP_URL_PATH );
+
+	// Bail if the current request is not to an uploaded image.
+	if ( substr( $_SERVER['REQUEST_URI'], 0, strlen( $baseurl ) ) !== $baseurl ) {
+		return;
+	}
+
+	$matches = array();
+	$relpath = str_replace( $baseurl, '', $_SERVER['REQUEST_URI'] );
+
+	// Bail if the current request is not to a subsize.
+	if ( ! preg_match( '#/(.+)-(\d+)x(\d+)\.([a-zA-Z]+)$#', $relpath, $matches ) ) {
+		return;
+	}
+
+	$img_width  = intval( $matches[2] );
+	$img_height = intval( $matches[3] );
+
+	$size       = '';
+	$size_names = get_intermediate_image_sizes();
+	$sizes      = wp_get_additional_image_sizes();
+
+	// Find the requested subsize name.
+	foreach ( $size_names as $size_name ) {
+		$width = intval(
+			isset( $sizes[ $size_name ] )
+				? $sizes[ $size_name ]['width']
+				: get_option( "{$size_name}_size_w" )
+		);
+
+		$height = intval(
+			isset( $sizes[ $size_name ] )
+				? $sizes[ $size_name ]['height']
+				: get_option( "{$size_name}_size_h" )
+		);
+
+		// Try to normalize dimensions if only one part matches for the current size.
+		if ( $width > 0 && $img_width !== $width && $img_height === $height ) {
+			$dimensions = wp_constrain_dimensions( $width, $height, 0, $img_height );
+			$img_width  = $dimensions[0];
+		} elseif ( $img_width === $width && $img_height !== $height && $height > 0 ) {
+			$dimensions = wp_constrain_dimensions( $width, $height, $img_width, 0 );
+			$img_height = $dimensions[1];
+		}
+
+		if (
+			// If the current size exactly matches requested width and height.
+			( $width === $img_width && $height === $img_height ) ||
+			// If the current size matches the requested height and has a variable width.
+			( 0 === $width && $height === $img_height ) ||
+			// If the current size matches the requested width and has a variable height.
+			( $width === $img_width && 0 === $height )
+		) {
+			$size = $size_name;
+
+			// Stop finding the size name if the current size exactly matches requested width and height.
+			// If the current size has a variable width or height, we need to continue finding in case there
+			// is an exact match.
+			if ( $width && $height ) {
+				break;
+			}
+		}
+	}
+
+	// Bail if the subsize name is not found.
+	if ( empty( $size ) ) {
+		return;
+	}
+
+	// Bail if we can't find an attachment with requested file.
+	$attachment_id = attachment_url_to_postid( "{$uploads_dir['baseurl']}/{$matches[1]}.{$matches[4]}" );
```

Comment:
Would be great to use `path_join` here instead, that would help with creating a variable that can be used in this function a bit down.

```suggestion
    $path = path_join( $uploads_dir['baseurl'], "{$matches[1]}.{$matches[4]}" );
	$attachment_id = attachment_url_to_postid( $path );
```

Reply:
It won't help because those are two different values, in the first one we have `baseurl` whilst in the second one we use `basedir`. So, we can't use one variable for both.
https://github.com/WordPress/performance/blob/3a2ad73e4b190aca24c46ca0485a46f540715825/modules/images/webp-uploads/generation.php#L90 https://github.com/WordPress/performance/blob/3a2ad73e4b190aca24c46ca0485a46f540715825/modules/images/webp-uploads/generation.php#L106